File: tools/tool-browser-request/src/main/java/com/exceptioncoder/toolbox/browserrequest/service/CurlParser.java
package com.exceptioncoder.toolbox.browserrequest.service;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * 把浏览器复制的 curl 命令解析成 {method, url, headers, body}。
 * 覆盖 Chrome DevTools「Copy as cURL (bash)」/「Copy as cURL (cmd)」/ PowerShell 风格的常用形态：
 *   - -X / --request 指定方法（缺省时若有 body 自动为 POST，否则 GET）
 *   - -H / --header 头
 *   - -b / --cookie 注入 Cookie 头
 *   - --data / --data-raw / --data-binary / -d 作为请求体
 *   - 行尾续行符 `\` `^` 反引号会先做规整
 *   - 单引号 / 双引号 / 反斜杠转义
 *
 * 不实现 multipart -F / -F @file，超出粘贴回放的常见用途。
 */
public final class CurlParser {

    private CurlParser() {}

    public record ParsedCurl(String method, String url,
                             Map<String, String> headers, String body) {}

    public static ParsedCurl parse(String raw) {
        if (raw == null) throw new IllegalArgumentException("curl 文本为空");
        String normalized = raw.replace("\r\n", "\n")
                .replace("\\\n", " ")
                .replace("^\n", " ")
                .replace("`\n", " ")
                .trim();
        List<String> tokens = tokenize(normalized);
        if (tokens.isEmpty() || !"curl".equalsIgnoreCase(tokens.get(0))) {
            throw new IllegalArgumentException("不是有效的 curl 命令，应以 curl 开头");
        }

        String method = null;
        String url = null;
        String body = null;
        Map<String, String> headers = new LinkedHashMap<>();
        List<String> cookies = new ArrayList<>();

        for (int i = 1; i < tokens.size(); i++) {
            String t = tokens.get(i);
            switch (t) {
                case "-X", "--request" -> method = tokens.get(++i);
                case "-H", "--header" -> {
                    String h = tokens.get(++i);
                    int colon = h.indexOf(':');
                    if (colon > 0) {
                        String k = h.substring(0, colon).trim();
                        String v = h.substring(colon + 1).trim();
                        headers.put(k, v);
                    }
                }
                case "-b", "--cookie" -> cookies.add(tokens.get(++i));
                case "--data", "--data-raw", "--data-binary", "--data-urlencode", "-d" -> {
                    String d = tokens.get(++i);
                    body = (body == null) ? d : body + "&" + d;
                }
                case "--compressed", "-L", "--location", "-k", "--insecure",
                     "-i", "--include", "-s", "--silent", "-v", "--verbose",
                     "-O", "--remote-name" -> { /* 无值/无意义 flag，忽略 */ }
                case "-A", "--user-agent" -> headers.put("User-Agent", tokens.get(++i));
                case "-e", "--referer" -> headers.put("Referer", tokens.get(++i));
                case "-u", "--user" -> {
                    String creds = tokens.get(++i);
                    headers.put("Authorization", "Basic " +
                            java.util.Base64.getEncoder().encodeToString(creds.getBytes()));
                }
                default -> {
                    if (t.startsWith("-")) {
                        // 未识别的带值短/长选项：保守跳过下一个 token 以免错位
                        if (i + 1 < tokens.size() && !tokens.get(i + 1).startsWith("-")
                                && url == null) {
                            // 不吞 URL，留给下一轮判断
                        } else if (i + 1 < tokens.size() && !tokens.get(i + 1).startsWith("-")) {
                            i++;
                        }
                    } else if (url == null) {
                        url = t;
                    }
                }
            }
        }

        if (url == null) throw new IllegalArgumentException("curl 命令缺少 URL");
        if (!cookies.isEmpty()) {
            String joined = String.join("; ", cookies);
            headers.merge("Cookie", joined, (a, b) -> a + "; " + b);
        }
        if (method == null) method = (body != null) ? "POST" : "GET";
        return new ParsedCurl(method.toUpperCase(), url, headers, body);
    }

    /** 简易 shell tokenizer：识别 ' " 和 \ 转义。 */
    private static List<String> tokenize(String s) {
        List<String> out = new ArrayList<>();
        StringBuilder cur = new StringBuilder();
        boolean inSingle = false, inDouble = false, hasContent = false;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (inSingle) {
                if (c == '\'') { inSingle = false; }
                else { cur.append(c); }
                hasContent = true;
            } else if (inDouble) {
                if (c == '\\' && i + 1 < s.length()) {
                    char n = s.charAt(i + 1);
                    if (n == '"' || n == '\\' || n == '$' || n == '`') { cur.append(n); i++; }
                    else { cur.append(c); }
                } else if (c == '"') { inDouble = false; }
                else { cur.append(c); }
                hasContent = true;
            } else {
                if (c == '\'') { inSingle = true; hasContent = true; }
                else if (c == '"') { inDouble = true; hasContent = true; }
                else if (c == '\\' && i + 1 < s.length()) { cur.append(s.charAt(++i)); hasContent = true; }
                else if (Character.isWhitespace(c)) {
                    if (hasContent) { out.add(cur.toString()); cur.setLength(0); hasContent = false; }
                } else { cur.append(c); hasContent = true; }
            }
        }
        if (hasContent) out.add(cur.toString());
        return out;
    }
}

File: frontend/src/features/browser-request/utils/curlParser.ts
/**
 * 前端版 cURL 解析器，保持与后端 {@code CurlParser.java} 行为一致。
 * 支持：
 *   -X / --request  方法
 *   -H / --header   头
 *   -b / --cookie   Cookie 头
 *   -d / --data / --data-raw / --data-binary / --data-urlencode  请求体
 *   -A / --user-agent / -e / --referer / -u / --user (basic auth)
 *   行尾续行符 \ ^ ` 规整为空格
 *   单引号 / 双引号 / 反斜杠转义
 *
 * 不支持：multipart -F、@file（同后端）。
 * 解析失败抛 Error（不返回 null）。
 */

export interface ParsedCurl {
  method: string
  url: string
  headers: Record<string, string>
  body: string | undefined
}

export function parseCurl(raw: string): ParsedCurl {
  if (!raw || !raw.trim()) throw new Error('cURL 文本为空')
  const normalized = raw
    .replace(/\r\n/g, '\n')
    .replace(/\\\n/g, ' ')
    .replace(/\^\n/g, ' ')
    .replace(/`\n/g, ' ')
    .trim()
  const tokens = tokenize(normalized)
  if (tokens.length === 0 || tokens[0].toLowerCase() !== 'curl') {
    throw new Error('不是有效的 cURL 命令，应以 curl 开头')
  }

  let method: string | null = null
  let url: string | null = null
  let body: string | undefined
  const headers: Record<string, string> = {}
  const cookies: string[] = []

  for (let i = 1; i < tokens.length; i++) {
    const t = tokens[i]
    switch (t) {
      case '-X':
      case '--request':
        method = tokens[++i]
        break
      case '-H':
      case '--header': {
        const h = tokens[++i]
        const colon = h.indexOf(':')
        if (colon > 0) {
          headers[h.slice(0, colon).trim()] = h.slice(colon + 1).trim()
        }
        break
      }
      case '-b':
      case '--cookie':
        cookies.push(tokens[++i])
        break
      case '-d':
      case '--data':
      case '--data-raw':
      case '--data-binary':
      case '--data-urlencode': {
        const d = tokens[++i]
        body = body == null ? d : body + '&' + d
        break
      }
      case '--compressed':
      case '-L': case '--location':
      case '-k': case '--insecure':
      case '-i': case '--include':
      case '-s': case '--silent':
      case '-v': case '--verbose':
      case '-O': case '--remote-name':
        break
      case '-A': case '--user-agent':
        headers['User-Agent'] = tokens[++i]
        break
      case '-e': case '--referer':
        headers['Referer'] = tokens[++i]
        break
      case '-u': case '--user': {
        const creds = tokens[++i]
        headers['Authorization'] = 'Basic ' + btoa(creds)
        break
      }
      default:
        if (t.startsWith('-')) {
          // 未识别的带值/无值选项保守跳过下一个 token（避免吞 URL）
          if (i + 1 < tokens.length && !tokens[i + 1].startsWith('-') && url == null) {
            // 不动，让下轮判断
          } else if (i + 1 < tokens.length && !tokens[i + 1].startsWith('-')) {
            i++
          }
        } else if (url == null) {
          url = t
        }
    }
  }

  if (!url) throw new Error('cURL 命令缺少 URL')
  if (cookies.length > 0) {
    const joined = cookies.join('; ')
    headers['Cookie'] = headers['Cookie'] ? headers['Cookie'] + '; ' + joined : joined
  }
  if (method == null) method = body != null ? 'POST' : 'GET'
  return { method: method.toUpperCase(), url, headers, body }
}

/** 简易 shell tokenizer：识别单/双引号 + 反斜杠转义。 */
function tokenize(s: string): string[] {
  const out: string[] = []
  let cur = ''
  let inSingle = false, inDouble = false, hasContent = false
  for (let i = 0; i < s.length; i++) {
    const c = s[i]
    if (inSingle) {
      if (c === "'") inSingle = false
      else cur += c
      hasContent = true
    } else if (inDouble) {
      if (c === '\\' && i + 1 < s.length) {
        const n = s[i + 1]
        if (n === '"' || n === '\\' || n === '$' || n === '`') { cur += n; i++ }
        else cur += c
      } else if (c === '"') inDouble = false
      else cur += c
      hasContent = true
    } else {
      if (c === "'") { inSingle = true; hasContent = true }
      else if (c === '"') { inDouble = true; hasContent = true }
      else if (c === '\\' && i + 1 < s.length) { cur += s[++i]; hasContent = true }
      else if (/\s/.test(c)) {
        if (hasContent) { out.push(cur); cur = ''; hasContent = false }
      } else { cur += c; hasContent = true }
    }
  }
  if (hasContent) out.push(cur)
  return out
}
